WIP: [MULTIARCH-4995] ABI documentation updates for 4.17 (#82213)
openshift/openshift-docs: modules/installing-ocp-agent-ibm-z.adoc

[NOTE]
====
Currently, ISO boot support on {ibm-z-name} (`s390x``) is available only for {op-system-base-full} KVM, which provides the flexibility to choose either PXE or ISO-based installation. For installations with z/VM and Logical Partition (LPAR), only PXE boot is supported.
====

[id="networking-reqs-ibmz_{context}"]
== Networking requirements for {ibm-z-title}

> In {ibm-z-title} environments, advanced networking technologies such as Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent-Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z-title}.

Reviewer: Replace Original Storage Adapter with Open Systems Adapter

The following table lists the network devices that are supported for the network configuration override functionality:

[cols="3,2,2,2,2", options="header"]
|====
| Network device 
| z/VM  
| KVM 
| LPAR Classic 
| LPAR Dynamic Partition Manager (DPM)

> | Original Storage Architecture (OSA) virtual switch 

Reviewer: is this vswitch or OSA or both ?
Just use OSA here since Open System Adapter is explained above

| Not needed
| Not applicable
| Not needed
| Not needed

| Direct attached OSA 
| Supported
| Supported only through a Linux bridge 
| Supported
| Not needed

| RDMA over Converged Ethernet (RoCE) 
| Not needed

> | Only through a Linux bridge 

Reviewer: in line 46 the sentence is supported ony through a Linux bridge, is this the same here at line 52 and 58?
When reading this does it mean the ai.ip_cfg_override can be used in that context for KVM ?

[id="networking-reqs-ibmz_{context}"]
== Networking requirements for {ibm-z-title}

In {ibm-z-title} environments, advanced networking technologies such as Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent-Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z-title}.

The following table lists the network devices that are supported for the network configuration override functionality:

[cols="3,2,2,2,2", options="header"]
|====
| Network device 
| z/VM  
| KVM 
| LPAR Classic 
| LPAR Dynamic Partition Manager (DPM)

| Original Storage Architecture (OSA) virtual switch 
| Not needed
| Not applicable
| Not needed
| Not needed

| Direct attached OSA 
| Supported
| Only through a Linux bridge 
| Supported
|  Not needed

| RDMA over Converged Ethernet (RoCE) 
| Not needed
| Only through a Linux bridge 
| Not needed
| Not needed

| HiperSockets 
| Supported 
| Only through a Linux bridge 
| Supported 
| Not needed 
|====

> // Configuring network overrides in {ibm-z-title}
> include:modules/configuring-network-overrides-ibmz.adoc[leveloffset=+2]

Reviewer 2: This needs to be included in the assembly, not at the bottom of this module. You'd want to place it under [this line](https://github.com/openshift/openshift-docs/blob/main/installing/installing_with_agent_based_installer/prepare-pxe-assets-agent.adoc?plain=1#L39).

Additionally, the include needs an extra colon:
```suggestion
// Configuring network overrides in {ibm-z-title}
include::modules/configuring-network-overrides-ibmz.adoc[leveloffset=+2]
```

[NOTE]
====
Currently, ISO boot support on {ibm-z-name} (`s390x``) is available only for {op-system-base-full} KVM, which provides the flexibility to choose either PXE or ISO-based installation. For installations with z/VM and Logical Partition (LPAR), only PXE boot is supported.
====

[id="networking-reqs-ibmz_{context}"]
==  Networking requirements for {ibm-z-title}

> In {ibm-z} environments, advanced networking technologies like Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z}.

Reviewer 2: It seems that just {ibm-z} does not resolve to anything. Looking at our [attributes file](https://github.com/openshift/openshift-docs/blob/main/_attributes/common-attributes.adoc#L269-L270), you need to choose either {ibm-z-name} for the product title with the copyright icon or {ibm-z-title} for the title without it:
```suggestion
In {ibm-z-title} environments, advanced networking technologies like Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z-title}.
```

> In {ibm-z} environments, advanced networking technologies like Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z}.

Reviewer 2: Per the [ISG word usage](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#l) entry for "like", it should not be used in place of "such as":
```suggestion
In {ibm-z} environments, advanced networking technologies such as Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z}.
```

> In {ibm-z} environments, advanced networking technologies like Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z}.

Reviewer 2: ```suggestion
In {ibm-z} environments, advanced networking technologies like Original Storage Architecture (OSA), HiperSockets, and Remote Direct Memory Access (RDMA) over Converged Ethernet (RoCE) require specific configurations that deviate from the standard settings used in the Agent-Based Installer deployments. These overrides are necessary to accommodate their unique requirements and ensure a successful and efficient deployment on {ibm-z}.
```

The following table lists the network devices that are supported for the network configuration override functionality:

[cols="3,2,2,2,2", options="header"]
|====
| Network device 
| z/VM  
| KVM 
| LPAR Classic 
| LPAR Dynamic Partition Manager (DPM)

| Original Storage Architecture (OSA) virtual switch 
| Not needed

> | --

Reviewer 2: Is this intentionally a dash, or should this say not required?

| Not needed
| Not needed

| Direct attached OSA 
| Supported

> | Only through a Linux bridge 

Reviewer 2: Would this still mean the same thing if it said "Supported only through a Linux bridge"? If so, I feel like that would be clear for all these entries that say "Only through a Linux bridge"

| Supported

> | Not required

Reviewer 2: Is there a reason this cell says not required and the others say not needed? If not, please make them consistent (I would personally say that not required sounds better)

| RDMA over Converged Ethernet (RoCE) 
| Not needed
| Only through a Linux bridge 
| Not needed
| Not needed

| HiperSockets 
| Supported 
| Only through a Linux bridge 
| Supported 
| Not needed 
|====

[id="configuring-network-overrides-ibmz_{context}"]

> === Configuring network overrides in {ibm-z-title}

Reviewer 2: This seems like the beginning of a procedure. It was somewhat acceptable to include the networking requirements section from above in this module, because it was additional conceptual information being added to a concept module. 

But this section, if it's meant to give users instructions for performing a task, can't really be added to a module in the same way per our mod docs standards. This section should probably be added to its own procedural module and then included in the assembly.

> You can specify a static IP address on {ibm-z} machines that uses Logical Partition (LPAR) and z/VM. This is specially useful when the network devices do not have a static MAC address assigned to them. 

Reviewer 2: This attribute also doesn't resolve in the preview:
```suggestion
You can specify a static IP address on {ibm-z-title} machines that uses Logical Partition (LPAR) and z/VM. This is specially useful when the network devices do not have a static MAC address assigned to them. 
```

> You can specify a static IP address on {ibm-z} machines that uses Logical Partition (LPAR) and z/VM. This is specially useful when the network devices do not have a static MAC address assigned to them. 

Reviewer 2: I think specially should be dropped here, but if you want to keep it, it should probably say "especially" instead:
```suggestion
You can specify a static IP address on {ibm-z} machines that uses Logical Partition (LPAR) and z/VM. This is useful when the network devices do not have a static MAC address assigned to them. 
```

> If you have an existing `.parm` file, edit it to include the following entry:
> 
> [source,terminal]
> ----
> ai.ip_cfg_override=1
> ----

Reviewer 2: If this overall section is moved to a procedure module, this should be made into a single step procedure:
```suggestion
.Procedure

* If you have an existing `.parm` file, edit it to include the following entry:
+
[source,terminal]
----
ai.ip_cfg_override=1
----
```

This parameter allows the file to add the network settings to the CoreOS installer.

> .Example of the `.parm` file

Reviewer 2: This can be trimmed down:
```suggestion
.Example`.parm` file
```

[source,terminal]
----
rd.neednet=1 cio_ignore=all,!condev
console=ttysclp0
coreos.live.rootfs_url=<coreos_url> <1>
ip=<ip>::<gateway>:<netmask>:<hostname>::none nameserver=<dns> 
rd.znet=qeth,<network_adaptor_range>,layer2=1
rd.<disk_type>=<adapter> <2> 
rd.zfcp=<adapter>,<wwpn>,<lun> random.trust_cpu=on <3>
zfcp.allow_lun_scan=0 
ai.ip_cfg_override=1
ignition.firstboot ignition.platform.id=metal 
random.trust_cpu=on
----
<1> For the `coreos.live.rootfs_url` artifact, specify the matching `rootfs` artifact for the `kernel` and `initramfs` that you are booting. Only HTTP and HTTPS protocols are supported.
<2> For installations on direct access storage devices (DASD) type disks, use `rd.` to specify the DASD where {rhel-first} is to be installed. For installations on Fibre Channel Protocol (FCP) disks, use `rd.zfcp=<adapter>,<wwpn>,<lun>` to specify the FCP disk where {rhel} is to be installed.

> <3> Specify values for `adapter`, `wwpn`, and `lun` as in the following example: `rd.zfcp=0.0.8002,0x500507630400d1e3,0x4000404600000000`.

Reviewer 2: So there are a few things about this snippet that need addressing:

1. If this is supposed to be an example file, then the user replaced values such as <coreos_url> should be an actual value instead of a placeholder. If you cannot find example values in time for 4.17 GA, please make sure you address this in another PR as soon as you can.
2. The callout labels below the snippet don't seem appropriate to include for a step about adding the `ai.ip_cfg_override=1` parameter to the file, since this procedure does not talk about configuring things like the `coreos.live.rootfs_url` artifact and `rd.<disk_type>=<adapter>`. You should probably remove the callout labels from the snippet and remove these descriptions.
